provider/cluster/manager.go

```diff
@@ -254,16 +285,72 @@ func (dm *deploymentManager) startTeardown() <-chan error {
 	return dm.do(dm.doTeardown)
 }
 
-func (dm *deploymentManager) doDeploy() error {
+func (dm *deploymentManager) doDeploy() ([]string, error) {
+	var err error
+
+	allHostnames := util.AllHostnamesOfManifestGroup(*dm.mgroup)
+	// Either reserve the hostnames, or confirm that they already are held
+	reservationResult := dm.hostnameService.ReserveHostnames(allHostnames, dm.lease)
+	withheldHostnames, err := reservationResult.Wait(dm.lc.ShuttingDown())
+	if err != nil {
+		deploymentCounter.WithLabelValues("reserve-hostnames", "err").Inc()
+		dm.log.Error("deploy hostname reservation error", "state", dm.state, "err", err)
+		return nil, err
+	}
+	deploymentCounter.WithLabelValues("reserve-hostnames", "success").Inc()
+
+	dm.log.Info("hostnames withheld", "cnt", len(withheldHostnames))
+
 	// Don't use a context tied to the lifecycle, as we don't want to cancel Kubernetes operations
 	ctx := context.Background()
-	result := dm.client.Deploy(ctx, dm.lease, dm.mgroup)
+
+	err = dm.client.Deploy(ctx, dm.lease, dm.mgroup)
 	label := "success"
-	if result != nil {
+	if err != nil {
 		label = "fail"
 	}
 	deploymentCounter.WithLabelValues("deploy", label).Inc()
-	return result
+
+	// Figure out what hostnames to declare
+	blockedHostnames := make(map[string]struct{})
+	for _, hostname := range withheldHostnames {
+		blockedHostnames[hostname] = struct{}{}
+	}
+	hosts := make(map[string]manifest.ServiceExpose)
+	hostToServiceName := make(map[string]string)
+	for _, service := range dm.mgroup.Services {
+		for _, expose := range service.Expose {
+			if !util.ShouldBeIngress(expose) {
+				continue
+			}
+
+			if dm.config.DeploymentIngressStaticHosts {
+				uid := clusterutil.IngressHost(dm.lease, service.Name)
+				host := fmt.Sprintf("%s.%s", uid, dm.config.DeploymentIngressDomain)
+				hosts[host] = expose
+				hostToServiceName[host] = service.Name
+			}
+
+			for _, host := range expose.Hosts {
+				_, blocked := blockedHostnames[host]
+				if !blocked {
+					hosts[host] = expose
+					hostToServiceName[host] = service.Name
+				}
+			}
+		}
+	}
+
+	for host, serviceExpose := range hosts {
+		err = dm.client.DeclareHostname(ctx, dm.lease, host, hostToServiceName[host], uint32(serviceExpose.ExternalPort))
+		if err != nil {
+			// TODO - counter
+			return withheldHostnames, err
```

Comment:
should it unreserve undeclared hostnames?

Reply:
I'm not sure what you're asking? How would this reserve something that isn't declared?

Reply:
I might have read this wrong.  I thought it was:

1. reserving all hostnames [further above](https://github.com/ovrclk/akash/pull/1349/files/13529347b830af3c213901d93c94a3ecbdaf9069#diff-d0fa560a89bd5c31e83e29b07f984194896f8f4604cfceb60d02a3e9c10c30f6R293)
2. "declaring" the hostnames (I assumed creating the ingress plumbing configuration) here.

So if it's reserving then declaring, but declaring doesn't complete, we might want to unreserve the undeclared hostnames.  I was probably reading it wrong.

Reply:
What you're describing is correct.

If this fails, the deployment fails. At that time the goroutine in `run()` same file has a `defer` that calls `dm.hostnameService.ReleaseHostnames(dm.lease)` . So it should drop any hostnames reserved at that point.